tasks/win.ps1

```diff
@@ -0,0 +1,22 @@
+[CmdletBinding()]
+Param(
+  [String]$message = "",
+  [Int]$timeout = 3
+)
+# If an error is encountered, the script will stop instead of the default of "Continue"
+$ErrorActionPreference = "Stop"
+
+If (Test-Path -Path $env:SYSTEMROOT\sysnative\shutdown.exe) {
+  $executable = "$env:SYSTEMROOT\sysnative\shutdown.exe"
+}
+ElseIf (Test-Path -Path $env:SYSTEMROOT\system32\shutdown.exe) {
+  $executable = "$env:SYSTEMROOT\system32\shutdown.exe"
+}
+Else {
+  $executable = "shutdown.exe"
+}
+
+cmd /c start $executable /r /t $timeout /d p:4:1 $message
+
+Write-Output "{""status"":""queued"",""timeout"":${timeout}}"
```

Comment:
So because the reboot command on line 19 is a blocking command, I think there is no chance that this line will ever be reached, as the server will reboot in the meantime, or has testing showed otherwise?

Reply:
Right you are, is the process that is spawned a child of the process running this script? I think it would be... That scenario would expose us to a race condition where the `Write-Output` completes before the `shutdown.exe` has registered the shutdown, the parent process then exits, (possibly?) killing all its children. The best way to test would be add a few seconds wait in the child process so that you can be sure it doesn't get killed when the parent exits

Reply:
As in test with a multi-second timeout? The task exits immediately after writing output, and the reboot still seems to happen.

Reply:
The thing is it's not relevant time the timeout, regardless of what the timeout is, `shutdown.exe` need to register the reboot with the system. If it gets killed before it does that it will never get rebooted. Since these are in two different threads there is a chance that one simply gets process faster than another, leading possibly to weird issues. The way to test would be to replace:

```powershell
cmd /c start /b $executable /r /t $timeout /d p:4:1
```

```powershell
ping 127.0.0.1 -n 6 > nul & cmd /c start /b $executable /r /t $timeout /d p:4:1
```

And see if the reboot still occurs. The ping will slow down the process containing the reboot to show properly if it is currently working due to luck or if the race condition truly doesn't exist. I tested this on my environment when I originally wrote the task and it was an issue but I don't have that test environment anymore

Reply:
I haven't been able to get the actual task to fail. I think that the session doesn't actually exit until shutdown is invoked (we don't force a disconnect, we wait for the target to declare the command complete). It continues to work with `cmd /c start /b /wait ...`, or if I directly invoke `shutdown.exe`. So switched to directly invoking shutdown.

Reply:
If you can't get it to fail then Im happy